pkg/daemon/daemon.go

```diff
@@ -379,39 +382,68 @@ func (dn *Daemon) restartDevicePluginPod() error {
 	dn.mu.Lock()
 	defer dn.mu.Unlock()
 	glog.V(2).Infof("restartDevicePluginPod(): try to restart device plugin pod")
-	pods, err := dn.kubeClient.CoreV1().Pods(namespace).List(metav1.ListOptions{
-		LabelSelector: "app=sriov-device-plugin",
-		FieldSelector: "spec.nodeName=" + dn.name,
-	})
-	if err != nil {
-		glog.Warningf("restartDevicePluginPod(): Failed to list device plugin pod: %s", err)
+
+	var podName string
+	if err := wait.PollImmediateUntil(3*time.Second, func() (bool, error) {
+		pods, err := dn.kubeClient.CoreV1().Pods(namespace).List(metav1.ListOptions{
+			LabelSelector: "app=sriov-device-plugin",
+			FieldSelector: "spec.nodeName=" + dn.name,
+		})
+
+		if errors.IsNotFound(err) {
+			glog.Info("restartDevicePluginPod(): device plugin pod exited")
+			return true, nil
+		}
+
+		if err != nil {
+			glog.Warningf("restartDevicePluginPod(): Failed to list device plugin pod: %s, retrying", err)
+			return false, nil
+		}
+
+		if len(pods.Items) == 0 {
+			glog.Info("restartDevicePluginPod(): device plugin pod exited")
+			return true, nil
+		}
+		podName = pods.Items[0].Name
+		return true, nil
+	}, dn.stopCh); err != nil {
+		glog.Errorf("restartDevicePluginPod(): failed to wait for pod deletion: %v", err)
+		return err
+	}
+
+	if podName == "" {
+		glog.Info("restartDevicePluginPod(): device plugin pod was not there")
 		return nil
 	}
-	if len(pods.Items) > 0 {
-		glog.V(2).Infof("restartDevicePluginPod(): Found device plugin pod %s", pods.Items[0].GetName())
-		err = dn.kubeClient.CoreV1().Pods(namespace).Delete(pods.Items[0].GetName(), &metav1.DeleteOptions{})
+
+	if err := wait.PollImmediateUntil(3*time.Second, func() (bool, error) {
+		glog.V(2).Infof("restartDevicePluginPod(): Found device plugin pod %s, deleting it", podName)
```

Comment:
what happens if `len(pods.Items) == 0`? Will it cause config daemon exit?

Reply:
You mean if there are no pods when we reach the delete? I am not sure if deleting a pod that does not exist returns an error. In that case the behaviour would be even worse (infinite loop).

This should not happen given the current flow (apart from external factors), but let me check if the error is notfound

Reply:
In previous version, it checks `pods.items >0` before deleting the pod.
With current change, there is no conditional check. I'm trying to understand why it's removed.

Reply:
Ah, that's because if `len (pods.Items) == 0` we don't find a pod to delete and we exit in line 414.

Reply:
ack, I missed the pfName checking. sorry for the noise.